### DISCUSSION TITLE
OS Error: 2 while opening a file

### DISCUSSION THREAD

**Discussion by Redstonetill** [pybricks/support#1745] [Q&A] [answered]
I want to open a file (self.file = open(filename, "rb")) on my pybricks Spike Prime robot, but i am getting an OSError: 2 and the errorcodes ({1: 'EPERM', 5: 'EIO', 16: 'EBUSY', 19: 'ENODEV', 22: 'EINVAL', 95: 'EOPNOTSUPP', 11: 'EAGAIN', 116: 'ETIMEDOUT', 140: 'ECANCELED'}) doesnt include this error. So I am wondering what this error means since I didnt find anything. (I am using pybricksdev to upload files in VS Code)
=> This is my VS Code configuration
{
            "name": "Launch File",
            "type": "debugpy",
            "request": "launch",
            "module": "pybricksdev",
            "args": ["run", "ble", "${file}"],
        },

**Comment by dlech** [comment]
Pybricks for SPIKE Prime doesn't have a file system currently, so it isn't possible to create a file.

If you want to save a small amount of data, like a calibration value, you can use `hub.system.storage`. See https://docs.pybricks.com/en/stable/hubs/primehub.html#pybricks.hubs.PrimeHub.system.storage

Or if you want log data, you can use the `print()` function and save the data on your computer.

**Reply by Redstonetill** [reply]
I have a 7MB file that I want to access at specific indexes (thats how it fits into the 250KB ram of the robot). So I dont think that I will fit inside of hub.system.storage. But I am wondering then why the documentation says, that their is a way to access files: 
![grafik](https://github.com/user-attachments/assets/43cc2d4a-7347-4a60-ad85-662f5bea4520)

**Reply by dlech** [reply]
On EV3, it is possible to have files, but on other hubs it isn't possible.

Looks like we could do a better job indicating that in the docs.

**Reply by Redstonetill** [reply]
Ok, is there any over way for me to do it?

**Reply by dlech** [reply]
Not sure, other than fetching small chunks at a time via Bluetooth.  Can you describe what the data is and what you want to do with it? Maybe there is some way to compress it or calculate it on the fly?

**Reply by Redstonetill** [reply]
Their isn't any way to calculate it, because it is rgb data which I need for the color sensors. But if I save it as an JPG image, it only takes up 47kB. But I dont know if there is a way to load an JPG Image partially and after that delete the partially loaded image. Since I am taking part in an competition it isn't possible for me to use bluetooth

**Reply by dlech** [reply]
> Their isn't any way to calculate it, because it is rgb data which I need for the color sensors

How many color sensors do you have than need 7MB of data? :open_mouth: 


> But I dont know if there is a way to load an JPG Image partially and after that delete the partially loaded image

There is no JPEG decoder library on the SPIKE Prime, so not really an option. 

I don't think it is enabled, but we might be able to consider gzip decompression support.


> And out of some reason using del actually doesnt unload it from memory:

This is because `del` only removes the name from the globals dict. It does not free the object that it used to point to.

If there are no other references to that object, the object will be freed the next time the garbage collector runs.

**Reply by Redstonetill** [reply]
> How many color sensors do you have than need 7MB of data? 😮

I am driving around the playing field and i am calculating with the wheel rotations where I am currently on the map. But this isnt perfect, so I am trying to use the Colorsensors to determine my position better. For that I have 7MB of rgb data (just field data). 
I have also thought about things like cutting out parts that I am not going to drive on or just simply storing grayscale values or using an lower resolution. But I am trying to fit the whole map at the current resolution into the robot because I dont know right now where the robot is going to drive or how accurate the rgb data has to be

**Reply by Redstonetill** [reply]
But is there any way to run the garbage collector using code? Since this would solve the entire problem I am having

**Reply by dlech** [reply, marked as answer]
Yes, there is a `gc` module.

```python
import gc

...

gc.collect()
```

There is also a `gc.threshold(()` function that can be useful instead of manually calling `gc.collect()`.

https://docs.micropython.org/en/latest/library/gc.html

**Reply by Redstonetill** [reply]
Really a big thank you, the memory has been freed up when I test it. Really big thanks
